src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -70,13 +70,27 @@ bool VMOperationTimeoutTask::is_armed() {
   return Atomic::load_acquire(&_armed) != 0;
 }
 
-void VMOperationTimeoutTask::arm() {
+void VMOperationTimeoutTask::arm(const char* vm_op_name) {
+  _vm_op_name = vm_op_name;
   _arm_time = os::javaTimeNanos();
   Atomic::release_store_fence(&_armed, 1);
 }
 
 void VMOperationTimeoutTask::disarm() {
   Atomic::release_store_fence(&_armed, 0);
+
+  // The two stores to `_armed` are counted in VM-op, but they should be
+  // insignificant compared to the actual VM-op duration.
+  jlong vm_op_duration = nanos_to_millis(os::javaTimeNanos() - _arm_time);
+
+  // Repeat the timeout-check logic on the VM thread, because
+  // VMOperationTimeoutTask might miss the arm-disarm window depending on
+  // the scheduling.
+  assert(Thread::current()->is_VM_thread(), "Check timeout on VM thread");
+  if (vm_op_duration > AbortVMOnVMOperationTimeoutDelay) {
+    fatal("%s VM operation took too long: completed in " JLONG_FORMAT " ms (timeout: " INTX_FORMAT " ms)",
+          _vm_op_name, vm_op_duration, AbortVMOnVMOperationTimeoutDelay);
+  }
```

Comment:
Should we null `_vm_op_name` here?

Reply:
It's not read by anyone from here on, so setting it to null is not needed, and a waste of the VM thread time.

Reply:
My concern would be holding the pointer to something that is already unallocated long after VM op completes. I guess it is a lower risk for `const char*`, which is probably a reference to globally available string, but anyway. The overhead argument does not seem very relevant here, as we do other stuff in this handler when it is enabled. It is not a waste in default VM mode.

Reply:
I see; cleared `_vm_op_name` at the end of `disarm()`.